File: src/integrations/redisBus.js
export class RedisBus {
  constructor(config) {
    this.config = config;
    this.client = null;
    this.enabled = Boolean(config.redis.enabled && config.redis.url);
    this.status = this.enabled ? "connecting" : "disabled";
    this.error = "";
  }

  async start() {
    if (!this.enabled || this.client) return;
    try {
      const module = await import("ioredis");
      const Redis = module.default || module.Redis;
      this.client = new Redis(this.config.redis.url, {
        lazyConnect: true,
        maxRetriesPerRequest: 1,
        enableOfflineQueue: false
      });
      this.client.on("error", (error) => {
        this.status = "error";
        this.error = error.message;
      });
      this.client.on("connect", () => {
        this.status = "connected";
        this.error = "";
      });
      await this.client.connect();
    } catch (error) {
      this.status = "unavailable";
      this.error = error.message;
      this.client = null;
    }
  }

  async publish(topic, payload) {
    if (!this.client || this.status === "disabled") return false;
    try {
      await this.client.publish(`${this.config.redis.namespace}:${topic}`, JSON.stringify(payload));
      return true;
    } catch (error) {
      this.status = "error";
      this.error = error.message;
      return false;
    }
  }

  snapshot() {
    return {
      enabled: this.enabled,
      status: this.status,
      namespace: this.config.redis.namespace,
      error: this.error
    };
  }
}

File: backend/app/integrations/redis_bus.py
from __future__ import annotations

import json

from backend.app.core.config import Settings


class RedisBus:
    def __init__(self, settings: Settings):
        self.settings = settings
        self.client = None
        self.enabled = bool(settings.redis_enabled and settings.redis_url)
        self.status = "disabled" if not self.enabled else "connecting"
        self.error = ""

    async def start(self) -> None:
        if not self.enabled:
            return
        try:
            import redis.asyncio as redis

            self.client = redis.from_url(self.settings.redis_url, decode_responses=True)
            await self.client.ping()
            self.status = "connected"
            self.error = ""
        except Exception as exc:  # pragma: no cover - depends on optional Redis
            self.client = None
            self.status = "unavailable"
            self.error = str(exc)

    async def publish(self, topic: str, payload: dict) -> bool:
        if not self.client:
            return False
        try:
            await self.client.publish(f"{self.settings.redis_namespace}:{topic}", json.dumps(payload, default=str))
            return True
        except Exception as exc:  # pragma: no cover - depends on optional Redis
            self.status = "error"
            self.error = str(exc)
            return False

    def snapshot(self) -> dict:
        return {
            "enabled": self.enabled,
            "status": self.status,
            "namespace": self.settings.redis_namespace,
            "error": self.error,
        }
